### PR TITLE
Fix new trait with all as supertraits sug name clash

### LINKED ISSUE #104635: New trait with all other traits as supertrait suggestion never take into account user provided trait names
Given the following code: [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=5d6f2aa16ca42efb503a4a9e2525e625)

```rust
trait NewTrait {}
trait NewTrait2 {}

fn foo() -> NewTrait + NewTrait2 {}

fn main() {}
```

The current output is:

```
Compiling playground v0.0.1 (/playground)
error[[E0782]](https://doc.rust-lang.org/stable/error-index.html#E0782): trait objects must include the `dyn` keyword
 --> src/main.rs:4:13
  |
4 | fn foo() -> NewTrait + NewTrait2 {}
  |             ^^^^^^^^^^^^^^^^^^^^
  |
help: add `dyn` keyword before this trait
  |
4 | fn foo() -> dyn NewTrait + NewTrait2 {}
  |             +++

error[[E0225]](https://doc.rust-lang.org/stable/error-index.html#E0225): only auto traits can be used as additional traits in a trait object
 --> src/main.rs:4:24
  |
4 | fn foo() -> NewTrait + NewTrait2 {}
  |             --------   ^^^^^^^^^ additional non-auto trait
  |             |
  |             first non-auto trait
  |
  = help: consider creating a new trait with all of these as supertraits and using that trait here instead: `trait NewTrait: NewTrait + NewTrait2 {}`
  = note: auto-traits like `Send` and `Sync` are traits that have special properties; for more information on them, visit <https://doc.rust-lang.org/reference/special-types-and-traits.html#auto-traits>

Some errors have detailed explanations: E0225, E0782.
For more information about an error, try `rustc --explain E0225`.
error: could not compile `playground` due to 2 previous errors
```

Ideally the suggestion should avoid suggesting a trait with the same name as the user provided traits.

### PR DESCRIPTION
**Pull request by Rageking8**
Fixes #104635

### CODE REVIEW

File: compiler/rustc_hir_analysis/src/astconv/mod.rs [hunk 1]
```diff
@@ -1312,14 +1312,24 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
...
+                .collect::<Vec<_>>();
+            let sug_trait_name = if regular_traits_vec.contains(&"NewTrait".to_string()) {
+                let mut append_no = 2;
+                while regular_traits_vec.contains(&format!("NewTrait{append_no}")) {
+                    append_no += 1;
+                }
+                format!("NewTrait{append_no}")
+            } else {
+                "NewTrait".to_string()
+            };
```
Comment (poliorcetics) [inline compiler/rustc_hir_analysis/src/astconv/mod.rs:1327]:
```suggestion
            let mut regular_traits = regular_traits
                .iter()
                .map(|t| t.trait_ref().print_only_trait_path().to_string());
            let sug_trait_name = regular_traits
                .filter_map(|name| name.strip_prefix("NewTrait")?.parse::<u32>::().ok())
                .max()
                .map_or(|| std::cow::Cow::from("NewTrait", |max| std::cow::Cow::from(format!("NewTrait{}", max + 1)));
```

Several allocations removed :)

**Comment by compiler-errors** [comment]
I'm a bit swamped

r? compiler

**Review by jackh726 — COMMENTED** [review]
LGTM, other than the other review :)

**Comment by Rageking8** [comment]
> LGTM, other than the other review :)

Would you prefer the current one or the one in the other review?

**Comment by Rageking8** [comment]
Maybe you have time now?
@compiler-errors

PR is a little stale atm.
Just need feedback on whether to adopt the suggested code by the other reviewer and a final PR approval.

**Comment by jackh726** [comment]
Waiting for CI to be fixed.

**Comment by Rageking8** [comment]
@jackh726 Since I will not have access to my PC for close to a week and I have a bit of trouble fixing the error, could you take over if you have spare time? Thanks.

**Pull request closed without merging**